File: backend/_examples/python-service/mongo_service.py
"""
MongoDB database configuration and connection management.

This module handles MongoDB connection pooling using module-level variables
to reuse connections across Lambda invocations, improving performance and reducing
cold start time.
"""

from pymongo import MongoClient

# Module-level MongoDB client for connection pooling across Lambda invocations
# Persists between invocations within the same Lambda container
MONGO_CLIENT = None

def get_mongo_version(config):
    """
    Retrieves the MongoDB version using connection pooling.

    Reuses the module-level MONGO_CLIENT across invocations for better performance.
    Executes the admin.buildInfo command to retrieve version information.
    On connection failure or command error, resets the client to None and raises the error.

    Connection pooling strategy:
    - First invocation: Creates a new MongoDB client and stores it in MONGO_CLIENT
    - Subsequent invocations: Reuses the existing client
    - On error: Resets MONGO_CLIENT to None to force reconnection on next invocation

    Returns:
        str: The MongoDB version string, or "unknown" if retrieval fails

    Raises:
        Exception: If connection or command fails
    """
    global MONGO_CLIENT
    try:
        # Create client if not already pooled
        if MONGO_CLIENT is None:
            MONGO_CLIENT = MongoClient(**config)

        # Execute buildInfo command and extract version
        info = MONGO_CLIENT.admin.command("buildInfo")
        return info.get("version", "unknown")
    except Exception as e:
        # Connection or command failed - reset client and log error
        print("MongoDB error: %s", str(e))
        MONGO_CLIENT = None
        raise

File: backend/_examples/nodejs-service/mongo-service.js
/**
 * MongoDB database configuration and connection management.
 *
 * This module handles MongoDB connection pooling using module-level variables
 * to reuse connections across Lambda invocations, improving performance and reducing
 * cold start time.
 */

const { MongoClient } = require("mongodb");

/**
 * Module-level MongoDB client for connection pooling across Lambda invocations.
 * Persists between invocations within the same Lambda container.
 * @type {MongoClient|null}
 */
let mongoClient = null;

/**
 * Retrieves the MongoDB version using connection pooling.
 *
 * Reuses the module-level mongoClient across invocations for better performance.
 * Executes the admin.buildInfo command to retrieve version information.
 * On connection failure or command error, resets the client to null and throws the error.
 *
 * Connection pooling strategy:
 * - First invocation: Creates a new MongoDB client and stores it in mongoClient
 * - Subsequent invocations: Reuses the existing client
 * - On error: Resets mongoClient to null to force reconnection on next invocation
 *
 * @param {*} config MongoDB configuration object with host, port, user, password
 * @returns {Promise<string>} The MongoDB version string, or "unknown" if retrieval fails
 * @throws {Error} If connection or command fails
 */
async function getMongoVersion(config) {
  try {
    // Build MongoDB connection URI — include credentials only when provided
    const auth = config.user ? `${config.user}:${config.password}@` : "";
    const uri = `mongodb://${auth}${config.host}:${config.port}`;

    // Create client if not already pooled
    if (!mongoClient) {
      mongoClient = new MongoClient(uri, {
        serverSelectionTimeoutMS: config.serverSelectionTimeoutMS,
        socketTimeoutMS: config.socketTimeoutMS,
        ...(config.isLocal ? {} : { tls: true, tlsAllowInvalidCertificates: true, retryWrites: false }),
      });
      await mongoClient.connect();
    }

    // Execute buildInfo command and extract version
    const info = await mongoClient.db("admin").command({ buildInfo: 1 });
    return info.version || "unknown";
  } catch (error) {
    // Connection or command failed - reset client and throw error
    console.error("MongoDB error:", error.message);
    mongoClient = null;
    throw error;
  }
}

module.exports = {
  getMongoVersion,
};
